File: golang-for-nodejs-developers/examples/async_await.go
package main

import (
	"errors"
	"fmt"
	"log"
	"time"
)

func hello(name string) chan any {
	ch := make(chan any, 1)
	go func() {
		time.Sleep(1 * time.Second)
		if name == "fail" {
			ch <- errors.New("failed")
		} else {
			ch <- "hello " + name
		}
		close(ch)
	}()

	return ch
}

func main() {
	result := <-hello("bob")
	switch v := result.(type) {
	case string:
		fmt.Println(v)
	case error:
		log.Println(v)
	}

	result = <-hello("fail")
	switch v := result.(type) {
	case string:
		fmt.Println(v)
	case error:
		log.Println(v.Error())
	}
}

File: golang-for-nodejs-developers/examples/async_await.js
import { setTimeout } from 'node:timers/promises';

async function hello(name) {
  await setTimeout(1000);
  if (name === 'fail') {
    throw new Error('failed');
  }
  return `hello ${name}`;
}

async function main() {
  try {
    let output = await hello('bob');
    console.log(output);

    output = await hello('fail');
    console.log(output);
  } catch (err) {
    console.error(err.message);
  }
}

main();
